### DISCUSSION TITLE
sqlalchemy typing of float column

### DISCUSSION THREAD

**Discussion by xmatthias** [sqlalchemy/sqlalchemy#9307] [Usage Questions] [answered]
All typing examples in the docs seem to use String/Integer combinations - where types seem to work - but i've been unable to type a Float column correctly (so it's actually got a float type) except with the workaround from https://github.com/dropbox/sqlalchemy-stubs/issues/178#issuecomment-1048703611 - which i hope isn't the intended way to use sqlalchemy's float type - and i fear might have other downsides.

I'm using a very simple example here - but it shows my current problem / doubt quite nicely.

`pyproject.toml`:
``` yml
[tool.mypy]
plugins = [
  "sqlalchemy.ext.mypy.plugin"
]
```

``` python
from sqlalchemy import Column
from sqlalchemy import Integer, Numeric, String, select, Float
from sqlalchemy.orm import declarative_base

# "Base" is a class that is created dynamically from the
# declarative_base() function
Base = declarative_base()


class User(Base):
    __tablename__ = "user"

    id = Column(Integer, primary_key=True)
    name = Column(String)
    age = Column(Float)

some_user = User(id=5, name="user")

reveal_type(some_user)
reveal_type(some_user.id)
reveal_type(some_user.name)
reveal_type(some_user.age)
```

Using `sqlalchemy 2.0.3` and `mypy==1.0.0`, i get the following output from mypy

```
sqlalchemy_disc.py:21: note: Revealed type is "sqlalchemy_disc.User"
sqlalchemy_disc.py:22: note: Revealed type is "Union[builtins.int, None]"
sqlalchemy_disc.py:23: note: Revealed type is "Union[builtins.str, None]"
sqlalchemy_disc.py:24: note: Revealed type is "Union[_N`1, None]"
```

running mypy on this file reveals the type of `age` as `Union[_N`1, None]`.

Now this seems wrong - and i'd expect the type of `age` to be `Union[builtins.float, None]`

Explicitly typing age doesn't work and changes the problem to an error.

``` python
# ...
    age:float = Column(Float)
# ...
```

```
sqlalchemy_disc.py:17: error: [SQLAlchemy Mypy plugin] Left hand assignment 'age: "float"' not compatible with ORM mapped expression of type "Mapped[_N]"  [misc]
sqlalchemy_disc.py:21: note: Revealed type is "sqlalchemy_disc.User"
sqlalchemy_disc.py:22: note: Revealed type is "Union[builtins.int, None]"
sqlalchemy_disc.py:23: note: Revealed type is "Union[builtins.str, None]"
sqlalchemy_disc.py:24: note: Revealed type is "builtins.float"
```

**Comment by zzzeek** [comment, marked as answer]
are you using SQLAlchemy 2.0 or 1.4?  if 1.4, I have no additional solutions, as 1.4 is only marginally typed.   if on 2.0, modern [annotated form](https://docs.sqlalchemy.org/en/20/orm/declarative_tables.html#using-annotated-declarative-table-type-annotated-forms-for-mapped-column) should resolve any typing issues with floats

**Reply by xmatthias** [reply]
luckily SQLAlchemy 2.0.

Is the `Column()` syntax (not sure what the name for that is - the one used in the sample above) considered deprecated in favor of the `mapped_column()` syntax, or is this just a "difficult to properly support" case, which requires the switch to `mapped_column()` syntax?

**Reply by zzzeek** [reply]
typing is only supported with mapped_column() syntax.  it's not possible in an ORM context for Column to work with typing without any plugins.

migration is pretty much search and replace to start with, see https://docs.sqlalchemy.org/en/20/changelog/whatsnew_20.html#migrating-an-existing-mapping

**Reply by xmatthias** [reply]
Perfect, thanks for your help!

great support :+1:
